### PR TITLE
[DSM] DDP-8349: lookup fields bug

### PR DESCRIPTION
**Pull request by gmakharat**
## Context

Ticket: https://broadinstitute.atlassian.net/browse/DDP-8349

From very beginning when `PatchRoute` refactored, we wrote the script which was doing export on every input change, but problem arose for look up type fields, look up type fields send several `NameValue` objects and triggering elastic export for each them was losing data, since our script was way faster for reading than elastic for writing values into it, most simplest and cleanest approach was to send only one ES request for several `NameValue` object.

**Things have done:** 

- Changed `GeneratorPayload` to hold collection of `NameValue` instead of single `NameValue`.
- Overloaded method in `BasePatch` which takes list of `NameValue` if we have several name value pairs.
- Used template method design pattern for methods `getElement` and `parseElement` to avoid duplication and added condition in this methods if we have several name value objects to behave differently.
- Added tests for `CollectionSourceGenerator` and `CollectionMappingGenerator` to test generate method.

**How to test:**

1. Find the participant which has either medical record or onc history tab. (You can easily filter participants by choosing `Duplicate` column of `Medical Record` and search for `Dupliacte` - `Yes`).
2. If it is medical record tab, open it and choose any medical record, it will transfer you to the medical record page, where you will see the input called `Confirmed Institution Name`, put some letter in this field and wait look up functionality to find appropriate options for you, choose any of it, then go back, reload the participant list, visit the same participant and check if chosen values are presented correctly.
3. If it is onc history tab, open it and you will see one or several onc histories, you can directly put values into columns without opening it, find column called `Facility` and do the same look up procedure like medical record case.

## Checklist

- [ ] I have labeled the type of changes involved using the `C-*` labels.
- [ ] I have assessed potential risks and labeled using the `R-*` labels.
- [ ] I have considered error handling and alerts, and added `L-*` labels as needed.
- [ ] I have considered security and privacy, and added `I-*` labels as needed
- [ ] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.
- [ ] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [ ] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc.

_If unsure or need help with any of the above items, add the `help wanted` label. For items that starts with `If applicable`, if it is not applicable, check it off and add `n/a` in front._

## FUD Score

_Overall, how are you feeling about these changes?_

- [ ] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [ ] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [ ] I have written automated positive tests
- [ ] I have written automated negative tests
- [ ] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [ ] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/patch/BasePatch.java [hunk 1]
```diff
@@ -116,6 +128,7 @@ Optional<Object> processSingleNameValue() {
 
     List<Object> processMultipleNameValues() {
         List<Object> updatedNameValues = new ArrayList<>();
+        exportToESWithId(patch.getId(), patch.getNameValues());
```
Comment (SimoneMaiwald) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/patch/BasePatch.java]: can we change the order here please. I would like to first write into db and then into ES.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/patch/OncHistoryDetailPatch.java [hunk 2]
```diff
@@ -95,7 +95,7 @@ Object handleSingleNameValue() {
     @Override
     Optional<Object> processEachNameValue(NameValue nameValue) {
         Patch.patch(oncHistoryDetailId, patch.getUser(), nameValue, dbElement);
-        exportToESWithId(oncHistoryDetailId, nameValue);
+        //exportToESWithId(oncHistoryDetailId, nameValue);
```
Comment (SimoneMaiwald) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/patch/OncHistoryDetailPatch.java]: shall we remove that, when it is now done per BasePatch

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/patch/ParticipantDataPatch.java [hunk 3]
```diff
@@ -52,7 +52,7 @@ Optional<Object> processEachNameValue(NameValue nameValue) {
         ParticipantDataNameValue participantDataNameValue =
                 new ParticipantDataNameValue(nameValue.getName(), nameValue.getValue(), ddpInstance.getDdpInstanceIdAsInt(),
                         patch.getDdpParticipantId(), patch.getFieldId());
-        exportToESWithId(participantDataId, participantDataNameValue);
+        //exportToESWithId(participantDataId, participantDataNameValue);
```
Comment (SimoneMaiwald) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/patch/ParticipantDataPatch.java]: shall we remove that, when it is now done per BasePatch

Comment (gmakharat) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/patch/ParticipantDataPatch.java]: good catch, totally forgot those comments.

**Review by ScalaChamp — APPROVED** [review]

**Review by SimoneMaiwald — APPROVED** [review]

**Pull request merged**